org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServiceAccessor.java

```diff
@@ -584,6 +583,32 @@ private static Collection<LanguageServerWrapper> getMatchingStartedWrappers(@Non
 		return serverInfos;
 	}
 
+
+	/**
+	 * Internal variant of getLanguageServers(IProject, Predicate<ServerCapabilities>, boolean)
+	 *
+	 * TODO: this doesn't replicate the behaviour with inactive LS: we return an async for the restarting server
+	 * that will only schedule further work after it completes. The older method calls wrapper.getServer() which
+	 * returns null on the event thread but blocks till completion on any other thread - does it matter?
+	 *
+	 * @param project
+	 * @param request
+	 * @param onlyActiveLS
+	 * @return
+	 */
+	@NonNull
+	static List<@NonNull CompletableFuture<LanguageServerWrapper>> getWrappers(@Nullable IProject project, Predicate<ServerCapabilities> request, boolean onlyActiveLS) {
+		List<@NonNull CompletableFuture<LanguageServerWrapper>> result = new ArrayList<>();
+		for (LanguageServerWrapper wrapper : startedServers) {
+			if ((!onlyActiveLS || wrapper.isActive()) && (project == null || wrapper.canOperate(project))) {
+				if (capabilitiesComply(wrapper, request)) {
```

Comment:
This part which is common between this method and getLanguageServers can be factored out. Then this method can go to `LanguageServers`. I will create a PR that we can merge standalone to make this PR smaller.

Reply:
Thanks @rubenporras

Reply:
You could remove this method now by rebasing and moving the specific part to LanguageServers. See my method in the LanguageServerExecutor in my PR:
 
```
		@Override
		protected @NonNull List<@NonNull CompletableFuture<@Nullable LanguageServerWrapper>> getServers() {
			// Compute list of servers from project & filter
			List<@NonNull CompletableFuture<LanguageServerWrapper>> wrappers = new ArrayList<>();
			for (LanguageServerWrapper wrapper :  LanguageServiceAccessor.getStartedWrappers(project, getFilter(), !restartStopped)) {
				wrappers.add(wrapper.getInitializedServer().thenApply(ls -> wrapper));
			}
			return wrappers;
		}
```

Reply:
Thanks - have done as you suggested.